REGRESSION (272869@main): Actions for QR codes sometimes fail to show up when long pressing

whsieh opened the pull request:
#### 6e1c1a40d3481d2c8207408e85fc4ccf6ad9ad72
<pre>
REGRESSION (272869@main): Actions for QR codes sometimes fail to show up when long pressing
<a href="https://bugs.webkit.org/show_bug.cgi?id=268889">https://bugs.webkit.org/show_bug.cgi?id=268889</a>
<a href="https://rdar.apple.com/122433291">rdar://122433291</a>

Reviewed by Megan Gardner.

Fix several issues in the dynamically-inserted image analysis menu item codepath below, following
the changes in 272869@main; importantly, this allows long-pressing ESIM and EID QR codes to show
&quot;Add eSIM&quot; menu items by default in WebKit apps, such as Mail. See below for more details:

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _insertDynamicImageAnalysisContextMenuItemsIfPossible]):

1.  Stop looking for the &quot;Look Up&quot; placeholder item. Prior to 272869@main, this hidden item was
    immediately inserted into the menu if image analysis results were not available upon long press,
    and only replaced with the final image analysis menu items in the case where the client didn&apos;t
    explicitly remove the Look Up item (`_WKElementActionTypeRevealImage`). However, the &quot;Look Up&quot;
    item may now be added immediately (in disabled state), which replaces the placeholder item and
    causes us to return early from `-_insertDynamicImageAnalysisContextMenuItemsIfPossible` due to
    not having a placeholder to replace, even if there are MRC menu items that need to be added.

    Fix this by changing the `indexOfPlaceholderItem` to a single boolean flag, always inserting the
    dynamic items at the end of the menu, and leave out the check for the `.hidden` menu item
    attribute, so that we&apos;ll only withhold the MRC items in the case where the client explicitly
    removed the &quot;Look Up&quot; item (which matches existing behavior).

2.  In a similar vein, it no longer makes sense to key the enablement of &quot;Look Up&quot; / &quot;Copy Subject&quot;
    off of whether or not the placeholder item exists, since the disabled menu item is already
    present; as such, we should remove this early return entirely and always enable these disabled
    items if necessary.

3.  Lastly, only attempt to dynamically add the &quot;Show Text&quot; action in the case where &quot;Show Text&quot;
    wasn&apos;t already in the menu, to prevent the &quot;Show Text&quot; action from unnecessarily moving within
    the menu in the case where OCR finishes quickly for an image with results.

Canonical link: <a href="https://commits.webkit.org/274222@main">https://commits.webkit.org/274222@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc824e5431f402a47ae709dd03cce7eb8c9e9ccd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17280 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40663 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40894 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34093 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14618 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14582 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12689 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42173 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/34883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38510 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/13257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36710 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14818 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8622 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14282 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

megangardner commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -12352,14 +12343,24 @@ - (void)_insertDynamicImageAnalysisContextMenuItemsIfPossible
-            addAction(_WKElementActionTypeImageExtraction);
+        if (!foundShowTextItem && self.hasSelectableTextForImageContextMenu) {
+            // Dynamically insert the "Show Text" menu item, if it wasn't already inserted.
+            // FIXME: This should probably be inserted unconditionally, and enabled if needed
```
Sorry if this is a stupid question, but why not just doit this way now? Too complicated?

whsieh replied:
(that, and too risky :/). Even if it works as intended, I think there's the latent risk of the context menu becoming filled with too many disabled items (3, including this "Show Text" action). It's possible that's not too bad in practice, but it's probably better to live on that change for a while...

Thanks for the review!

whsieh committed:
REGRESSION (272869@main): Actions for QR codes sometimes fail to show up when long pressing
https://bugs.webkit.org/show_bug.cgi?id=268889
rdar://122433291

Reviewed by Megan Gardner.

Fix several issues in the dynamically-inserted image analysis menu item codepath below, following
the changes in 272869@main; importantly, this allows long-pressing ESIM and EID QR codes to show
"Add eSIM" menu items by default in WebKit apps, such as Mail. See below for more details:

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _insertDynamicImageAnalysisContextMenuItemsIfPossible]):

1.  Stop looking for the "Look Up" placeholder item. Prior to 272869@main, this hidden item was
    immediately inserted into the menu if image analysis results were not available upon long press,
    and only replaced with the final image analysis menu items in the case where the client didn't
    explicitly remove the Look Up item (`_WKElementActionTypeRevealImage`). However, the "Look Up"
    item may now be added immediately (in disabled state), which replaces the placeholder item and
    causes us to return early from `-_insertDynamicImageAnalysisContextMenuItemsIfPossible` due to
    not having a placeholder to replace, even if there are MRC menu items that need to be added.

    Fix this by changing the `indexOfPlaceholderItem` to a single boolean flag, always inserting the
    dynamic items at the end of the menu, and leave out the check for the `.hidden` menu item
    attribute, so that we'll only withhold the MRC items in the case where the client explicitly
    removed the "Look Up" item (which matches existing behavior).

2.  In a similar vein, it no longer makes sense to key the enablement of "Look Up" / "Copy Subject"
    off of whether or not the placeholder item exists, since the disabled menu item is already
    present; as such, we should remove this early return entirely and always enable these disabled
    items if necessary.

3.  Lastly, only attempt to dynamically add the "Show Text" action in the case where "Show Text"
    wasn't already in the menu, to prevent the "Show Text" action from unnecessarily moving within
    the menu in the case where OCR finishes quickly for an image with results.

Canonical link: https://commits.webkit.org/274222@main
